core/src/main/scala/shapeless/smc.scala

```diff
@@ -263,67 +381,12 @@ class SMCOps[P, S <: SMC](p: P, val smc: S) {
   def reverse(implicit reverse: S#Reverse[P]): reverse.Out = reverse(p)
   def map(f: Poly)(implicit mapper: S#Mapper[f.type, P]): mapper.Out = mapper(p)
   def foldLeft[Z](z: Z)(f: Poly2)(implicit folder: S#LeftFolder[P, Z, f.type]): folder.Out = folder(p, z)
+
+  def rightAssoc(implicit rightAssoc: S#RightAssoc[P]): rightAssoc.Out = rightAssoc(p)
+  def leftAssoc(implicit leftAssoc: S#LeftAssoc[P]): leftAssoc.Out = leftAssoc(p)
 }
```

Comment:
A lot of this is to avoid calls of `swap`, yes? I guess that makes sense eventually at least from a performance PoV, but at this point it might be obscuring the essential structure.